From 7a901d9316b6f3e17a6f45caa094f5974cfeb8e4 Mon Sep 17 00:00:00 2001
From: Sveinung Kvilhaugsvik <sveinung84@users.sourceforge.net>
Date: Thu, 21 Jul 2016 20:05:26 +0200
Subject: [PATCH 3/3] Don't crash on savegame error reporting.

Say "missing" rather than trying to print the name of the missing extra the
activity target.

See bug #24868

---
 server/savegame2.c | 7 +++++--
 server/savegame3.c | 7 +++++--
 2 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/server/savegame2.c b/server/savegame2.c
index bd6d7ad..bd82c12 100644
--- a/server/savegame2.c
+++ b/server/savegame2.c
@@ -2680,8 +2680,11 @@ static void sg_load_players(struct loaddata *loading)
     unit_list_iterate(pplayer->units, punit) {
       if (!can_unit_continue_current_activity(punit)) {
         log_sg("Unit doing illegal activity in savegame!");
-        log_sg("Activity: %s, Target: %s", unit_activity_name(punit->activity),
-               extra_rule_name(punit->activity_target));
+        log_sg("Activity: %s, Target: %s",
+               unit_activity_name(punit->activity),
+               punit->activity_target ? extra_rule_name(
+                                          punit->activity_target)
+                                      : "missing");
         punit->activity = ACTIVITY_IDLE;
       }
     } unit_list_iterate_end;
diff --git a/server/savegame3.c b/server/savegame3.c
index a862314..2cac67a 100644
--- a/server/savegame3.c
+++ b/server/savegame3.c
@@ -3482,8 +3482,11 @@ static void sg_load_players(struct loaddata *loading)
     unit_list_iterate(pplayer->units, punit) {
       if (!can_unit_continue_current_activity(punit)) {
         log_sg("Unit doing illegal activity in savegame!");
-        log_sg("Activity: %s, Target: %s", unit_activity_name(punit->activity),
-               extra_rule_name(punit->activity_target));
+        log_sg("Activity: %s, Target: %s",
+               unit_activity_name(punit->activity),
+               punit->activity_target ? extra_rule_name(
+                                          punit->activity_target)
+                                      : "missing");
         punit->activity = ACTIVITY_IDLE;
       }
     } unit_list_iterate_end;
-- 
2.1.4

